types/block.go

```diff
@@ -1104,12 +1038,26 @@ type Data struct {
 
 // Hash returns the hash of the data
 func (data *Data) Hash() tmbytes.HexBytes {
-	if data == nil {
-		return (Txs{}).Hash()
+	if data.hash != nil {
+		return data.hash
 	}
-	if data.hash == nil {
-		data.hash = data.Txs.Hash() // NOTE: leaves of merkle tree are TxIDs
+
+	// compute the data availability header
+	// todo(evan): add the non redundant shares back into the header
+	shares, _ := data.ComputeShares()
+	rawShares := shares.RawShares()
+
+	squareSize := uint64(math.Sqrt(float64(len(shares))))
+
+	eds, err := da.ExtendShares(squareSize, rawShares)
+	if err != nil {
+		panic(err)
 	}
```

Comment:
we panicked before in [`fillDataAvailabilityHeader`](https://github.com/celestiaorg/celestia-core/blob/2d3768eb5f6ea6bb75cd0893f48de461ece4280a/types/block.go#L127), but I think this time it might be worse, as `ExtendShares` will return an error if the square size is too large!